🚨🚨🚨 [SuperPoint] Fix keypoint coordinate output and add post processing

sbucaille opened the pull request:
# What does this PR do?

There is a bug highlighted by @merveenoyan where SuperPoint returns absolute keypoint coordinates at a wrong scale.
For example, an image of size (1000, 1000) will first be resized into (640, 480) by the image processor but SuperPoint will output keypoint coordinates between 0 and 640 for x and 0 and 480 for y.
This PR fixes this bug by returning keypoints in relative coordinates out of the SuperPoint forward method.
An additional ``post_process_keypoint_detection`` is added to the ``SuperPointImageProcessor`` to cast the keypoint coordinates into the original image sizes.
I also added an "unwrapping" logic where the output will be a list of ``keypoints, scores, descriptors`` for each image so that the user don't have to deal with the current mask mechanism. This is conditionned by a boolean argument in the signature.

I have not finished all the details but most of the implementation is there and I will add tests tomorrow but at least the PR is open as I promised Merve ☺️

Fixes #33825


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. 
-> On Slack
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@amyeroberts @NielsRogge @merveenoyan

Commit:
feat: Added int conversion and unwrapping

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +273,28 @@ def preprocess(
 
         return BatchFeature(data=data, tensor_type=return_tensors)
+
+    def post_process_keypoint_detection(self, outputs, target_sizes, unwrap_batch_dim=True):
```
Could you add type hinting and a docstring for this please?

amyeroberts replied:
We should add this as a documented method for the model's doc page e.g. [like here ](https://github.com/huggingface/transformers/blob/c409cd81777fb27aadc043ed3d8339dbc020fb3b/docs/source/en/model_doc/deformable_detr.md?plain=1#L61)

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +273,28 @@ def preprocess(
+
+        # Convert masked_keypoints to int
+        masked_keypoints = outputs.keypoints.to(torch.int32)
+
+        outputs.keypoints = masked_keypoints
```
We shouldn't be modifying inputs - this is a side-effect and can cause surprising to users

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +273,28 @@ def preprocess(
+        outputs.keypoints = masked_keypoints
+
+        results = []
+        for i, image_mask in enumerate(outputs.mask):
+            indices = torch.nonzero(image_mask).squeeze(1)
+            keypoints = outputs.keypoints[i][indices]
+            scores = outputs.scores[i][indices]
+            descriptors = outputs.descriptors[i][indices]
+            results.append({"keypoints": keypoints, "scores": scores, "descriptors": descriptors})
```
it's more pythonic to iterate over objects directly 

```suggestion
        for (image_mask, keypoints, scores, descriptors) in zip(outputs.mask, outputs.keypoints, outputs.scores, outputs.descriptors):
            indices = torch.nonzero(image_mask).squeeze(1)
            keypoints = keypoints[indices]
            scores = scores[indices]
            descriptors = descriptors[indices]
            results.append({"keypoints": keypoints, "scores": scores, "descriptors": descriptors})
```

amyeroberts commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -258,6 +258,9 @@ def _extract_keypoints(self, scores: torch.Tensor) -> Tuple[torch.Tensor, torch.
         # Convert (y, x) to (x, y)
         keypoints = torch.flip(keypoints, [1]).float()
 
+        # Convert to relative coordinates
+        keypoints = keypoints / torch.tensor([width, height], device=keypoints.device)
```
Technically a breaking change - but as this is more of a fix, I think it's OK

amyeroberts reviewed (commented):
Thanks for fixing and adding this method! 

Overall looks good - just a few things to add / update

sbucaille committed:
test: added tests for post_process_keypoint_detection of SuperPointImageProcessor

sbucaille committed:
docs: changed docs to include post_process_keypoint_detection method and switched from opencv to matplotlib

sbucaille committed:
test: changed test to not depend on SuperPointModel forward

sbucaille committed:
test: added missing require_torch decorator

sbucaille committed:
docs: changed pyplot parameters for the keypoints to be more visible in the example

sbucaille committed:
tests: changed import torch location to make test_flax and test_tf

sbucaille committed:
Revert "tests: changed import torch location to make test_flax and test_tf"

This reverts commit 39b32a2f69500bc7af01715fc7beae2260549afe.

sbucaille committed:
tests: fixed import

NielsRogge commented on docs/source/en/model_doc/superpoint.md:
````diff
@@ -86,23 +86,28 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
 ```
 
-You can then print the keypoints on the image to visualize the result :
+You can then print the keypoints on the image of your choice to visualize the result :
````
```suggestion
You can then print the keypoints on the image of your choice to visualize the result:
```

NielsRogge commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -86,23 +86,28 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
-    color = tuple([score.item() * 255] * 3)
-    image = cv2.circle(image, (keypoint_x, keypoint_y), 2, color)
-cv2.imwrite("output_image.png", image)
+import matplotlib.pyplot as plt
+plt.axis("off")
```
```suggestion
import matplotlib.pyplot as plt

plt.axis("off")
```

merveenoyan commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +274,46 @@ def preprocess(
+        Args:
+            outputs ([`SuperPointKeypointDescriptionOutput`]):
+                Raw outputs of the model.
+            target_sizes (`torch.Tensor` of shape `(batch_size, 2)`):
```
small nit: I think we should also give an option to have this as a tuple of tuples of same size like in OWL to save users of the of converting it to tensor everytime and check inside whether it's a tensor or not
I'd imagine user passing in `[image.size for image in images]` the easiest

sbucaille replied:
You are right, I also thought about it in the first place. I've added this option

sbucaille committed:
chore: applied suggestions from code review

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

sbucaille committed:
tests: fixed import

sbucaille committed:
tests: fixed import (bis)

sbucaille committed:
tests: fixed import (ter)

sbucaille committed:
feat: added choice of type for target_size and changed tests accordingly

sbucaille committed:
docs: updated code snippet to reflect the addition of target size type choice in post process method

Commit:
tests: fixed imports (...)

Commit:
tests: fixed imports (...)

Commit:
style: formatting file

merveenoyan commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -86,23 +86,29 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
 
 inputs = processor(images, return_tensors="pt")
 outputs = model(**inputs)
+image_sizes = [(image[1], image[0]) for image in images]
```
it should be either `image.shape[1]` or `image.size[1]` I think and not directly image itself

it would've been directly to have `image.size` from convenience standpoint IMHO i.e. each tuple would've been `image.size` and not indexed

sbucaille replied:
Good catch, it's a typo from a wrong copy paste, should be fixed now. Also, the fact that it is `tuple(image.size[1], image.size[0])` is because we require `(h,w)` in input (I guess it's standard from what I've seen in other post process methods in the repo) but `image.size` returns `(w, h)`

merveenoyan commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -86,23 +86,29 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
+for output in outputs:
+    keypoints = output["keypoints"]
+    scores = output["scores"]
+    descriptors = output["descriptors"]
```
Can you add a line of example where one can use this too? They're used for matching mostly AFAIK but I don't know much :)

sbucaille committed:
docs: fixed typo from image[0] to image.size[0]

amyeroberts commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +277,54 @@ def preprocess(
+        masked_keypoints = outputs.keypoints.clone()
+
+        for keypoints, image_size in zip(masked_keypoints, image_sizes):
+            keypoints[:, 0] = keypoints[:, 0] * image_size[1]
+            keypoints[:, 1] = keypoints[:, 1] * image_size[0]
```
I think? Otherwise we're just taking the first and second elements of the batch. 

If I'm right - I'm surprised this wasn't caught. Has this been tested on batch inputs? 

```suggestion
            keypoints[:, 0] = keypoints[:, 0] * image_size[:, 1]
            keypoints[:, 1] = keypoints[:, 1] * image_size[:, 0]
```

sbucaille replied:
I don't think there is a problem with this part. After checking, `image_size` is indeed a `(2,)` tensor, `keypoints` as `(n, 2)`, so `keypoints[:, 0] = keypoints [:, 0] * image_size[1]` multiplies all elements x elements from `keypoints` with the scalar in `image_size`

amyeroberts commented on tests/models/superpoint/test_image_processing_superpoint.py:
```diff
@@ -110,3 +132,33 @@ def test_input_image_properly_converted_to_grayscale(self):
+                    keypoints[:, 1] <= image_size[0]
+                )
+                all_above_zero = torch.all(keypoints[:, 0] >= 0) and torch.all(keypoints[:, 1] >= 0)
+                self.assertTrue(all_below_image_size and all_above_zero)
```
it's better to assert one then another separately. Otherwise, if the test fails you have to re-run with a breakpoint to figure out what's happened

sbucaille replied:
Fixed

amyeroberts commented on tests/models/superpoint/test_image_processing_superpoint.py:
```diff
@@ -110,3 +132,33 @@ def test_input_image_properly_converted_to_grayscale(self):
         for image in pre_processed_images["pixel_values"]:
             self.assertTrue(np.all(image[0, ...] == image[1, ...]) and np.all(image[1, ...] == image[2, ...]))
+
+    @slow
```
Why mark as slow? It should be pretty fast

sbucaille replied:
This is a leftover from when I used SuperPoint model to produce the keypoint detection output. I removed it

amyeroberts commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -86,23 +86,29 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
+    s=scores * 50,
+    alpha=0.8
+)
+plt.savefig(f"output_image.png")
```
It would be nice to display the output image in the docs :)

sbucaille replied:
Image added !

amyeroberts reviewed (commented):
Thanks for iterating! 

Mostly small comments. Main one is about the indexing on `image_sizes` in the post-processing method

sbucaille committed:
docs: added output image and fixed some tests

sbucaille commented:
@amyeroberts comments addressed, tests are passing, let me know if I can do a final rebase for the merge

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33200). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel commented:
Hi @sbucaille, thanks for working on this and addressing comments, I will review this week!

qubvel commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -32,9 +32,16 @@
+if is_torch_available():
+    import torch
+
+    from .modeling_superpoint import SuperPointKeypointDescriptionOutput
```
In case it is used only for proper typehints
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .modeling_superpoint import SuperPointKeypointDescriptionOutput
```

sbucaille replied:
Ah yes you are right, it is indeed only used for typing, changed it

qubvel commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -258,6 +258,9 @@ def _extract_keypoints(self, scores: torch.Tensor) -> Tuple[torch.Tensor, torch.
         # Convert (y, x) to (x, y)
         keypoints = torch.flip(keypoints, [1]).float()
```
Do we have somewhere description of the keypoints output format? I mean it would be great to have it in docs and code snippets, and maybe mention it in `post_process_keypoint_detection` method

sbucaille replied:
The format is documented in `SuperPointKeypointDescriptionOutput` docstring, do you think I should still add this information somewhere else either way ?

qubvel reviewed (commented):
Thanks for working on this! Just some minor comments

sbucaille committed:
Update docs/source/en/model_doc/superpoint.md

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

sbucaille committed:
fix: included SuperPointKeypointDescriptionOutput in TYPE_CHECKING if statement and changed tests results to reflect changes to SuperPoint from absolute keypoints coordinates to relative

sbucaille committed:
docs: changed SuperPoint's docs to print output instead of just accessing

sbucaille commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -86,24 +86,31 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
+for output in outputs:
+    keypoints = output["keypoints"]
+    scores = output["scores"]
+    descriptors = output["descriptors"]
```
I've tweaked the docs a bit to add printing, at first I didn't really know what to do with this part because the most interesting one is below 😄

sbucaille committed:
style: applied make style

sbucaille commented:
@qubvel thanks for the review, I've addressed your comments. I initially forgot to change the tests to reflect the changes in terms of SuperPoint's outputs, so this is also fixed. Let me know if there is anything else to do

qubvel commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -260,7 +260,7 @@ def test_inference(self):
         inputs = preprocessor(images=images, return_tensors="pt").to(torch_device)
         with torch.no_grad():
             outputs = model(**inputs)
-        expected_number_keypoints_image0 = 567
```
Wondering why the number of keypoints changed :)

saeed6944 replied:
,

qubvel commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -275,11 +275,13 @@ def test_inference(self):
-        expected_keypoints_image0_values = torch.tensor([[480.0, 9.0], [494.0, 9.0], [489.0, 16.0]]).to(torch_device)
+        expected_keypoints_image0_values = torch.tensor([[0.75, 0.0188], [0.7719, 0.0188], [0.7641, 0.0333]]).to(
+            torch_device
+        )
         expected_scores_image0_values = torch.tensor(
-            [0.0064, 0.0137, 0.0589, 0.0723, 0.5166, 0.0174, 0.1515, 0.2054, 0.0334]
+            [0.0064, 0.0139, 0.0591, 0.0727, 0.5170, 0.0175, 0.1526, 0.2057, 0.0335]
         ).to(torch_device)
-        expected_descriptors_image0_value = torch.tensor(-0.1096).to(torch_device)
+        expected_descriptors_image0_value = torch.tensor(0.0449).to(torch_device)
```
Wondering why scores and descriptors also changed, do you have an idea?

sbucaille replied:
Ok this is weird, I tried to backtrack where a change could happen.
I rolled back to the point I checked out commit 56baa03380fc17d85705240ebbc57c075a9b3f23 which is from the original SuperPoint PR that passed the tests. At least was passing the tests, but it turns out I end up with the same error at this revision.
I don't know if it's something hardware, maybe you can try on your side and see if the tests are different between the two commits ?
I'm not available tonight but can have a look tomorrow morning

sbucaille replied:
Noticed that the absolute to relative coordinate conversion was done too early and had an impact on the descriptor values. So I moved it to the end of the model and the descriptor value is back to a "closer" value than the initial one, yet it's still slightly different from the initial integration test

qubvel reviewed (commented):
Thanks quick iteration, a few more comments

sbucaille committed:
docs: added missing output type and precision in docstring of post_process_keypoint_detection

sbucaille committed:
perf: deleted loop to perform keypoint conversion in one statement

sbucaille committed:
fix: moved keypoint conversion at the end of model forward

sbucaille committed:
docs: changed SuperPointInterestPointDecoder to SuperPointKeypointDecoder class name and added relative (x, y) coordinates information to its method

sbucaille commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +276,54 @@ def preprocess(
+
+        for keypoints, image_size in zip(masked_keypoints, image_sizes):
+            keypoints[:, 0] = keypoints[:, 0] * image_size[1]
+            keypoints[:, 1] = keypoints[:, 1] * image_size[0]
```
Changed that into a single statement

sbucaille commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +276,54 @@ def preprocess(
+        Args:
+            outputs ([`SuperPointKeypointDescriptionOutput`]):
+                Raw outputs of the model.
+            target_sizes (`torch.Tensor` or `List[Tuple[int, int]]`, *optional*):
```
Indeed, was initially optional in the first commits and forgot to change it, fixed

sbucaille commented on src/transformers/models/superpoint/image_processing_superpoint.py:
```diff
@@ -270,3 +276,54 @@ def preprocess(
+
+    def post_process_keypoint_detection(
+        self, outputs: "SuperPointKeypointDescriptionOutput", target_sizes: Union[TensorType, List[Tuple]]
+    ):
```
Output type hint added with relative keypoint coordinate information in the docstring

sbucaille committed:
fix: changed type hint

qubvel commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -485,6 +488,9 @@ def forward(
             mask[i, : _scores.shape[0]] = 1
 
+        # Convert to relative coordinates
+        keypoints[:, :] = keypoints[:, :] / torch.tensor([width, height], device=keypoints.device)
```
Do we need here `[:, :]`?

sbucaille replied:
Nope, fixed it

qubvel commented on tests/models/superpoint/test_modeling_superpoint.py:
```diff
@@ -275,11 +275,13 @@ def test_inference(self):
         expected_scores_image0_values = torch.tensor(
-            [0.0064, 0.0137, 0.0589, 0.0723, 0.5166, 0.0174, 0.1515, 0.2054, 0.0334]
+            [0.0064, 0.0139, 0.0591, 0.0727, 0.5170, 0.0175, 0.1526, 0.2057, 0.0335]
         ).to(torch_device)
-        expected_descriptors_image0_value = torch.tensor(-0.1096).to(torch_device)
+        expected_descriptors_image0_value = torch.tensor(-0.1095).to(torch_device)
```
Looks fine now! 👍

qubvel commented on src/transformers/models/superpoint/modeling_superpoint.py:
```diff
@@ -192,7 +192,7 @@ def forward(
 
 
-class SuperPointInterestPointDecoder(nn.Module):
+class SuperPointKeypointDecoder(nn.Module):
```
In case we rename it, it should be renamed across all the code/docs, but it's probably better to leave it with the original name, because it's out of scope for this PR (unless there is any strong reason)

sbucaille replied:
Not a necessary change, just by reading it again I thought it would be consistent to rename to this, but I'll open another PR for that matter, no problem

qubvel reviewed (commented):
Thank you for the fixes! Test diff looks fine now 👍

sbucaille committed:
refactor: removed unnecessary brackets

sbucaille committed:
revert: SuperPointKeypointDecoder to SuperPointInterestPointDecoder

qubvel commented on docs/source/en/model_doc/superpoint.md:
```diff
@@ -86,24 +86,32 @@ model = SuperPointForKeypointDetection.from_pretrained("magic-leap-community/sup
+import matplotlib.pyplot as plt
+
+plt.axis("off")
+plt.imshow(image)
+plt.scatter(
+    keypoints[:, 0],
+    keypoints[:, 1],
+    c=scores * 100,
+    s=scores * 50,
+    alpha=0.8
+)
```
```suggestion
plt.imshow(image_1)
plt.scatter(
    outputs[0]["keypoints"][:, 0],
    outputs[0]["keypoints"][:, 1],
    c=outputs[0]["scores"] * 100,
    s=outputs[0]["scores"] * 50,
    alpha=0.8
)
```

qubvel reviewed (commented):
Thanks, looks good to me now!

cc @ArthurZucker for final review

sbucaille committed:
Update docs/source/en/model_doc/superpoint.md

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

ArthurZucker reviewed (approved):
Feel free to merge if ready for you @qubvel 🤗

onuralpszr commented:
Hello, I was following this PR and curious that is it going to merge or any change still require ?

sbucaille commented:
Hi @onuralpszr ,
Thanks for asking, there is no change required on my side !

qubvel commented:
@sbucaille thanks a lot for your contribution!
